### DISCUSSION TITLE
CRL Delegate support

### DISCUSSION THREAD

**Discussion by Olegandr** [strongswan/strongswan#527] [General]
Hi, everyone
Per RFC 5280, Section 5,
> "CRL issuers issue CRLs.  The CRL issuer is either the CA or an entity that has been authorized by the CA to issue CRLs.  
>  CAs publish CRLs  to provide status information about the certificates they issued. 
> **However, a CA may delegate this responsibility to another trusted authority.**

Section 6.3.3, specifically step (f) describes how CRL chain of trust is validated:

>       (f)  Obtain and validate the certification path for the issuer of the complete CRL.  
>            The trust anchor for the certification path MUST be the same as the trust anchor used to validate the target certificate. 
>            If a key usage extension is present in the CRL issuer's certificate, verify that the cRLSign bit is set.

However when I am presenting StrongSwan (Linux strongSwan U5.8.2/K5.4.0-80-generic) with valid CRL, signed by the CRL delegate authority I receive error
```
Jul 28 22:17:51 testGW charon: 06[CFG] checking certificate status of "CN=Intermediate CA"
Jul 28 22:17:51 testGW charon: 06[CFG] ocsp check skipped, no ocsp found
Jul 28 22:17:51 testGW charon: 06[CFG]   fetching crl from 'http://10.2.3.3/CA_revocation.crl' ...
Jul 28 22:17:51 testGW charon: 06[CFG] issuer of fetched CRL 'CN=Intermediate CA' does not match certificate's issuer ' CN=Root CA'
Jul 28 22:17:51 testGW charon: 06[CFG] certificate status is not available
```
RFC 5280 does not say anything about comparing CRL issuer with certificate issuer. The check appears to be implemented incorrectly.

**Comment by tobiasbrunner** [comment]
If you use a separate CRL issuer, it has to be listed in the cRLIssuer field of a CDP in the end-entity certificate. See RFC 5280, section 4.2.1.13.:

    If the certificate issuer is not the CRL issuer, then the cRLIssuer
    field MUST be present and contain the Name of the CRL issuer.

Also see section 6.3.3. (b)(1) (note that strongSwan does not enforce the indirectCRL flag):

    If the DP includes cRLIssuer, then verify that the issuer
    field in the complete CRL matches cRLIssuer in the DP and
    that the complete CRL contains an issuing distribution
    point extension with the indirectCRL boolean asserted.
    Otherwise, verify that the CRL issuer matches the
    certificate issuer.

**Reply by Olegandr** [reply]
Oh, I see now. So will strongswan follow that logic if implemented correctly? 
This whole indirect CRS is a mess, looks like. It is not obvious how strongswan should obtain the CRL delegate certificate in the first place.

**Reply by tobiasbrunner** [reply]
> So will strongswan follow that logic if implemented correctly?

Yes, it accepts CRLs if the issuer either is the same as that of the checked certificate or it matches the cRLIssuer field in a CDP in that certificate.

> It is not obvious how strongswan should obtain the CRL delegate certificate in the first place.

It has to be pre-installed. I'm not aware of any standardized mechanism that allows fetching it from somewhere dynamically (such as CDPs for CRLs). Which is probably why this feature is not commonly used and CAs just sign their CRLs directly.

Maybe OCSP is an option for you as that makes delegation to a different entity easier because the OCSP response can contain the signer's certificate (chain).
